File: study/lua/src/app/scene/view/carnivalActivity/CarnivalActivityUIHelper.lua
local CarnivalActivityUIHelper = {}
local TypeConvertHelper = require("app.utils.TypeConvertHelper")
local DataConst = require("app.const.DataConst")
local UserDataHelper = require("app.utils.UserDataHelper")

function CarnivalActivityUIHelper._isSingleGoodExchange(actTaskUnitData)
    local fixRewards = actTaskUnitData:getRewardItems()
    local hasSelectRewards = actTaskUnitData:isSelectReward()
    return actTaskUnitData:isExchageType() and (not hasSelectRewards)
end

function CarnivalActivityUIHelper._checkRes(actTaskUnitData)
    local items = actTaskUnitData:getConsumeItems()
    dump(items)
    local LogicCheckHelper = require("app.utils.LogicCheckHelper")
    local canBuy = true
	for k,v in ipairs(items) do
		if v.type == TypeConvertHelper.TYPE_RESOURCE and v.value == DataConst.RES_JADE2 then
			canBuy = canBuy and LogicCheckHelper.enoughValue(v.type,v.value,v.size,true)
			if not canBuy then
				return canBuy
			end 
		else
			canBuy = canBuy and LogicCheckHelper.enoughValue(v.type,v.value,v.size,false)
			if not canBuy then
				G_Prompt:showTip(Lang.get("common_res_not_enough"))	
				return canBuy
			end  
		end  
    end
    return canBuy
end

function CarnivalActivityUIHelper._checkPkgFull(actTaskUnitData,index)
    local rewards = actTaskUnitData:getRewardItems()
    local selectRewards = actTaskUnitData:getSelectRewardItems()
    if selectRewards and selectRewards[index] then
        local newRewards = {}
        for k,v in ipairs(rewards) do
            table.insert( newRewards, v)
        end
        table.insert( newRewards, selectRewards[index])
        rewards = newRewards
    end
    local UserCheck = require("app.utils.logic.UserCheck")
    local full = UserCheck.checkPackFullByAwards(rewards)
    return  not full
end

function CarnivalActivityUIHelper._hasGold(actTaskUnitData)
    local items = actTaskUnitData:getConsumeItems()
    for k,v in ipairs(items) do
        if v.type == TypeConvertHelper.TYPE_RESOURCE or v.value == DataConst.RES_DIAMOND  then
            return v
        end
    end
    return nil
end

function CarnivalActivityUIHelper._showSelectRewarsPopup(actTaskUnitData,sendFunc)
	local function callBackFunction(awardItem, index, total)
		local function confirmFunction()
			if awardItem == nil then
				G_Prompt:showTip(Lang.get("recruit_choose_first"))
				return true
			end
			if not CarnivalActivityUIHelper._checkPkgFull(actTaskUnitData,index) then
				return
			end
			local rewardIndex = index

			if  actTaskUnitData:isActivityExpired() then
				G_Prompt:showTip(Lang.get("customactivity_tips_already_finish"))	
				return 
			end
			sendFunc(actTaskUnitData:getAct_id(),actTaskUnitData:getId(),rewardIndex,nil)
		end
		local goldItem = CarnivalActivityUIHelper._hasGold(actTaskUnitData) 
		if goldItem then
			local hintStr = Lang.get("customactivity_buy_confirm",{count = goldItem.size})
			UIPopupHelper.popupConfirm(hintStr,confirmFunction)
		else
			confirmFunction()
		end
	end

    local awardItems = actTaskUnitData:getSelectRewardItems()
    local PopupSelectReward = require("app.ui.PopupSelectReward").new(Lang.get("days7activity_receive_popup"),callBackFunction)
    PopupSelectReward:setTip(Lang.get("days7activity_receive_popup_tip"))
    PopupSelectReward:updateUI(awardItems)
    PopupSelectReward:openWithAction()
end

function CarnivalActivityUIHelper._getMaxBuyNum (consumeItems)
	local maxNum = 9999

	for k,v in ipairs(consumeItems) do 
		local ownNum = UserDataHelper.getNumByTypeAndValue(v.type, v.value)
		local newMaxNum = math.floor(ownNum / v.size)

		maxNum = maxNum > newMaxNum and newMaxNum or maxNum
	end

	return maxNum
end	

--活动批量兑换
function CarnivalActivityUIHelper.popupExchangeItems(actTaskUnitData,sendFunc)
	local LogicCheckHelper = require("app.utils.LogicCheckHelper")
	local value01,value02,onlyShowMax = actTaskUnitData:getProgressValue()
	local consumeItems = actTaskUnitData:getConsumeItems()
    local fixRewards = actTaskUnitData:getRewardItems()
    local selectRewards = actTaskUnitData:getSelectRewardItems()

	local surplus = value01-- 剩余购买次数
	local item = fixRewards[1]
	assert(item,"can not find exchange gain item")

	if surplus < 0 then
		return 
	end

	local function callBackFunction(exchangeItem,consumeItems,buyCount)
		local newConsumeItems = {}
		for k,v in ipairs(consumeItems) do 
			table.insert( newConsumeItems,{type = v.type,value = v.value,size = v.size*buyCount})
		end
		--是否能购买检测
		if LogicCheckHelper.enoughValueList(newConsumeItems, false) == false then
			G_Prompt:showTip(Lang.get("common_res_not_enough"))	
			return
		end
		
		if actTaskUnitData:isActivityExpired() then
			G_Prompt:showTip(Lang.get("customactivity_tips_already_finish"))	
			return 
		end

		sendFunc(actTaskUnitData:getAct_id(),actTaskUnitData:getId(),nil,buyCount)
	end

	--背包检测
	local isFull, leftCount = LogicCheckHelper.isPackFull(item.type,item.value)
	if isFull == true then
		return
	end

	local maxCanBuyNum = CarnivalActivityUIHelper._getMaxBuyNum(consumeItems)

	surplus = surplus > maxCanBuyNum and maxCanBuyNum or surplus

	-- 临时应对方案   一对多且剩余次数为多
	if surplus == 1 or ( surplus > 1 and #fixRewards > 1 ) then
		local function callbackOK()
			callBackFunction(item,consumeItems,1)
		end
		local function hasYuanBaoCost()
			-- body
			for i,value in ipairs(consumeItems) do
				if value.type == TypeConvertHelper.TYPE_RESOURCE and value.value == DataConst.RES_DIAMOND then
					return true
				end
			end
			return false
		end

		--只有元宝才弹提示
		if hasYuanBaoCost() then
			local title = Lang.get("custom_spentyuanbao_title")
			local content = Lang.get("custom_spentyuanbao_content", {num = consumeItems[1].size})
			local popupAlert = require("app.ui.PopupAlert").new(title, content, callbackOK)
			popupAlert:openWithAction()
		else
			callBackFunction(item,consumeItems,1)
		end

		return
	end

	local popup = require("app.ui.PopupItemExchange").new(Lang.get("common_title_exchange_item"),callBackFunction)
	popup:updateUI(item,consumeItems,surplus)
	popup:setMaxLimit(math.min(surplus,leftCount))
	popup:openWithAction()	
end

function CarnivalActivityUIHelper.receiveReward(actTaskUnitData,sendFunc)
	if not CarnivalActivityUIHelper._checkRes(actTaskUnitData) then
		return
	end

	if CarnivalActivityUIHelper._isSingleGoodExchange(actTaskUnitData) then
		CarnivalActivityUIHelper.popupExchangeItems(actTaskUnitData,sendFunc)
		return
	end

	if actTaskUnitData:isSelectReward() then--要弹窗，玩家选奖励　
		CarnivalActivityUIHelper._showSelectRewarsPopup(actTaskUnitData, sendFunc)
		return
	end

	-- 不是以物易物类型  但是条件达成可以兑换了   例如每日充值领取
	if CarnivalActivityUIHelper._checkPkgFull(actTaskUnitData) then 
		sendFunc(
				actTaskUnitData:getAct_id(),actTaskUnitData:getId(),nil,nil)
	end
end

return CarnivalActivityUIHelper

File: study/ts/script/scene/view/carnivalActivity/CarnivalActivityUIHelper.ts
import { LogicCheckHelper } from "../../../utils/LogicCheckHelper";
import { G_Prompt } from "../../../init";
import { Lang } from "../../../lang/Lang";
import { table } from "../../../utils/table";
import { UserCheck } from "../../../utils/logic/UserCheck";
import { TypeConvertHelper } from "../../../utils/TypeConvertHelper";
import { DataConst } from "../../../const/DataConst";
import { UIPopupHelper } from "../../../utils/UIPopupHelper";
import PopupBase from "../../../ui/PopupBase";
import PopupSelectReward from "../../../ui/PopupSelectReward";
import { assert } from "../../../utils/GlobleFunc";
import PopupAlert from "../../../ui/PopupAlert";
import PopupItemExchange from "../../../ui/PopupItemExchange";
import { UserDataHelper } from "../../../utils/data/UserDataHelper";

interface ICarnivalActivityUIHelper {
    _isSingleGoodExchange(actTaskUnitData): boolean;
    _checkRes(actTaskUnitData): boolean;
    _checkPkgFull(actTaskUnitData, index?): boolean;
    _hasGold(actTaskUnitData): any;
    _showSelectRewarsPopup(actTaskUnitData, sendFunc): void;
    popupExchangeItems(actTaskUnitData, sendFunc): void;
    receiveReward(actTaskUnitData, sendFunc): void;
    _getMaxBuyNum(consumeItems): number;
}
export var CarnivalActivityUIHelper = {} as ICarnivalActivityUIHelper;
CarnivalActivityUIHelper._isSingleGoodExchange = function (actTaskUnitData) {
    var fixRewards = actTaskUnitData.getRewardItems();
    var hasSelectRewards = actTaskUnitData.isSelectReward();
    return actTaskUnitData.isExchageType() && !hasSelectRewards;
};
CarnivalActivityUIHelper._checkRes = function (actTaskUnitData) {
    var items = actTaskUnitData.getConsumeItems();
    //dump(items);
    var canBuy = true;
    for (let k in items) {
        var v = items[k];
        if (v.type == TypeConvertHelper.TYPE_RESOURCE && v.value == DataConst.RES_JADE2) {
            canBuy = canBuy && LogicCheckHelper.enoughValue(v.type, v.value, v.size, true);
            if (!canBuy) {
                return canBuy;
            }
        } else {
            canBuy = canBuy && LogicCheckHelper.enoughValue(v.type, v.value, v.size, false);
            if (!canBuy) {
                G_Prompt.showTip(Lang.get('common_res_not_enough'));
                return canBuy;
            }
        }
    }
    return canBuy;
};
CarnivalActivityUIHelper._checkPkgFull = function (actTaskUnitData, index?) {
    var rewards = actTaskUnitData.getRewardItems();
    var selectRewards = actTaskUnitData.getSelectRewardItems();
    if (selectRewards && selectRewards[index]) {
        var newRewards = [];
        for (let k in rewards) {
            var v = rewards[k];
            table.insert(newRewards, v);
        }
        table.insert(newRewards, selectRewards[index]);
        rewards = newRewards;
    }
    var full = UserCheck.checkPackFullByAwards(rewards);
    return !full;
};
CarnivalActivityUIHelper._hasGold = function (actTaskUnitData) {
    var items = actTaskUnitData.getConsumeItems();
    for (let k in items) {
        var v = items[k];
        if (v.type == TypeConvertHelper.TYPE_RESOURCE || v.value == DataConst.RES_DIAMOND) {
            return v;
        }
    }
    return null;
};
CarnivalActivityUIHelper._showSelectRewarsPopup = function (actTaskUnitData, sendFunc) {
    function callBackFunction(awardItem, index, total) {
        function confirmFunction() {
            if (awardItem == null) {
                G_Prompt.showTip(Lang.get('recruit_choose_first'));
                return true;
            }
            if (!CarnivalActivityUIHelper._checkPkgFull(actTaskUnitData, index)) {
                return;
            }
            var rewardIndex = index + 1;
            if (actTaskUnitData.isActivityExpired()) {
                G_Prompt.showTip(Lang.get('customactivity_tips_already_finish'));
                return;
            }
            sendFunc(actTaskUnitData.getAct_id(), actTaskUnitData.getId(), rewardIndex, null);
        }
        var goldItem = CarnivalActivityUIHelper._hasGold(actTaskUnitData);
        if (goldItem) {
            var hintStr = Lang.get('customactivity_buy_confirm', { count: goldItem.size });
            UIPopupHelper.popupConfirm(hintStr, confirmFunction);
        } else {
            confirmFunction();
        }
    }
    var awardItems = actTaskUnitData.getSelectRewardItems();
    PopupBase.loadCommonPrefab('PopupSelectReward', (popup: PopupSelectReward) => {
        popup.ctor(Lang.get('days7activity_receive_popup'), callBackFunction);
        popup.setTip(Lang.get('days7activity_receive_popup_tip'));
        popup.updateUI(awardItems);
        popup.openWithAction();
    });

};
CarnivalActivityUIHelper._getMaxBuyNum = function (consumeItems) {
    var maxNum = 9999;
    for (var k in consumeItems) {
        var v = consumeItems[k];
        var ownNum = UserDataHelper.getNumByTypeAndValue(v.type, v.value);
        var newMaxNum = Math.floor(ownNum / v.size);
        maxNum = maxNum > newMaxNum && newMaxNum || maxNum;
    }
    return maxNum;
};
CarnivalActivityUIHelper.popupExchangeItems = function (actTaskUnitData, sendFunc) {
    var [value01, value02, onlyShowMax] = actTaskUnitData.getProgressValue();
    var consumeItems = actTaskUnitData.getConsumeItems();
    var fixRewards = actTaskUnitData.getRewardItems();
    var selectRewards = actTaskUnitData.getSelectRewardItems();
    var surplus = value01;
    var item = fixRewards[0];
    //assert((item, 'can not find exchange gain item');
    if (surplus < 0) {
        return;
    }
    function callBackFunction(exchangeItem, consumeItems, buyCount) {
        var newConsumeItems = [];
        for (let k in consumeItems) {
            var v = consumeItems[k];
            table.insert(newConsumeItems, {
                type: v.type,
                value: v.value,
                size: v.size * buyCount
            });
        }
        if (LogicCheckHelper.enoughValueList(newConsumeItems, false) == false) {
            G_Prompt.showTip(Lang.get('common_res_not_enough'));
            return;
        }
        if (actTaskUnitData.isActivityExpired()) {
            G_Prompt.showTip(Lang.get('customactivity_tips_already_finish'));
            return;
        }
        sendFunc(actTaskUnitData.getAct_id(), actTaskUnitData.getId(), null, buyCount);
    }
    var [isFull, leftCount] = LogicCheckHelper.isPackFull(item.type, item.value);
    if (isFull == true) {
        return;
    }
    var maxCanBuyNum = CarnivalActivityUIHelper._getMaxBuyNum(consumeItems);
    surplus = surplus > maxCanBuyNum && maxCanBuyNum || surplus;
    if (surplus == 1 || surplus > 1 && fixRewards.length > 1) {
        var callbackOK = function () {
            callBackFunction(item, consumeItems, 1);
        }
        var hasYuanBaoCost = function () {
            for (let i in consumeItems) {
                var value = consumeItems[i];
                if (value.type == TypeConvertHelper.TYPE_RESOURCE && value.value == DataConst.RES_DIAMOND) {
                    return true;
                }
            }
            return false;
        }
        if (hasYuanBaoCost()) {
            var title = Lang.get('custom_spentyuanbao_title');
            var content = Lang.get('custom_spentyuanbao_content', { num: consumeItems[0].size });
            PopupBase.loadCommonPrefab('PopupAlert', (popupAlert: PopupAlert) => {
                popupAlert.init(title, content, callbackOK);
                popupAlert.openWithAction();
            });
        } else {
            callBackFunction(item, consumeItems, 1);
        }
        return;
    }
    PopupBase.loadCommonPrefab('PopupItemExchange', (popup: PopupItemExchange) => {
        popup.ctor(Lang.get('common_title_exchange_item'), callBackFunction);
        popup.openWithAction();
        popup.updateUI(item, consumeItems, surplus);
        popup.setMaxLimit(Math.min(surplus, leftCount));

    });
};
CarnivalActivityUIHelper.receiveReward = function (actTaskUnitData, sendFunc) {
    if (!CarnivalActivityUIHelper._checkRes(actTaskUnitData)) {
        return;
    }
    if (CarnivalActivityUIHelper._isSingleGoodExchange(actTaskUnitData)) {
        CarnivalActivityUIHelper.popupExchangeItems(actTaskUnitData, sendFunc);
        return;
    }
    if (actTaskUnitData.isSelectReward()) {
        CarnivalActivityUIHelper._showSelectRewarsPopup(actTaskUnitData, sendFunc);
        return;
    }
    if (CarnivalActivityUIHelper._checkPkgFull(actTaskUnitData)) {
        sendFunc(actTaskUnitData.getAct_id(), actTaskUnitData.getId(), null, null);
    }
};
